### ISSUE TITLE
LSTM inference c++ threads block on DropoutState

### ISSUE THREAD

**Issue by deanyang11** [pytorch/pytorch#134637]
### 🐛 Describe the bug

I'd doing RL with c++ mutiple threads on a single GPU.
I find find threads are all block was block in lstm foward in get_dropout_state on a static varable:

RNN.cpp
_cudnn_impl(){
  ...
  auto& dropout_state = get_dropout_state(dropout_p, train, input.options());
  std::unique_lock<DropoutState> lock{dropout_state};
  ...
}

get_dropout_state(){
  ...
  static std::vector dropout_state_cache{
  static_cast<size_t>(cuda::getNumGPUs())};
  ...
}

And the DropoutState comments says:

// Every time we use a dropout state, we need to synchronize with its event,
// to make sure all previous uses finish running before this one starts. Once
// we’re done, we record the event to allow others to synchronize with this
// kernel. Those events are really needed only for inter-stream sync on a
// single GPU. I doubt anyone will want to run cuDNN RNNs in parallel on a
// single GPU, so they should end up being complete no-ops.

In fact lstm forwad doesn’t need a DropoutState, but it still mutex lock on it.
Is it possible to bypass the mutex lock in this situation?

### Versions

v2.4.0
